### PR TITLE
feat: Use POJOs to represent JSON objects read/written from wire

### PR DESCRIPTION
**Pull request by purplefox**
### Description 

Issue https://github.com/confluentinc/ksql/issues/4333

This PR introduces POJOs to represent JSON objects that we read/write from the wire as discussed on https://github.com/confluentinc/ksql/pull/4320

I'm not totally sold on this... it seems perhaps a little over-engineered for what we are doing, but tell me what you think anyway :)

### Testing done 

Includes new tests

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamArgs.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.Map;
+
+/**
+ * Represents the arguments to a query stream request
+ */
+public class QueryStreamArgs {
+
+  public final String sql;
+  public final Boolean push;
+  public final JsonObject properties;
```
Comment (big-andy-coates) [inline ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamArgs.java]: as above.

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamArgs.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
+  public QueryStreamArgs(final @JsonProperty(value = "sql", required = true) String sql,
+      final @JsonProperty(value = "push", required = true) Boolean push,
+      final @JsonProperty(value = "properties")
+          Map<String, Object> properties) {
```
Comment (big-andy-coates) [inline ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamArgs.java]: required `push` can be of type `boolean`.

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class QueryStreamResponseMetadata {
+
+  public final JsonArray columnNames;
+  public final JsonArray columnTypes;
+  public final String queryID;
+  public final Integer rowCount;
+
+  public QueryStreamResponseMetadata(final JsonArray columnNames, final JsonArray columnTypes,
+      final String queryID, final Integer rowCount) {
```
Comment (big-andy-coates) [inline ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java]: As above, prefer factory methods to having some fields that can be `null` and some that can't.

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    this.columnNames = columnNames;
+    this.columnTypes = columnTypes;
+    this.queryID = queryID;
+    this.rowCount = rowCount;
+  }
```
Comment (big-andy-coates) [inline ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java]: validate arguments.

File: ksql-api/src/main/java/io/confluent/ksql/api/server/AcksSubscriber.java [hunk 5]
```diff
@@ -31,9 +33,11 @@
  */
 public class AcksSubscriber extends ReactiveSubscriber<JsonObject> {
 
-  private static final Buffer ACK_RESPONSE_LINE = new JsonObject().put("status", "ok").toBuffer()
-      .appendString("\n");
   private static final Logger log = LoggerFactory.getLogger(AcksSubscriber.class);
+  private static final int BATCH_SIZE = 4;
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/AcksSubscriber.java]: out of date

File: ksql-api/src/main/java/io/confluent/ksql/api/server/ServerUtils.java [hunk 6]
```diff
@@ -37,30 +37,44 @@ private ServerUtils() {
...
-      final RoutingContext routingContext) {
-    try {
-      return new JsonObject(buffer);
-    } catch (DecodeException e) {
-      final String message = "Invalid JSON in request args";
-      log.error(message, e);
-      handleError(routingContext.response(), 400, ERROR_CODE_INVALID_JSON,
-          message);
-      return null;
+  public static <T> T deserialiseObject(final Buffer buffer, final HttpServerResponse response,
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: ack

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/PojoCodec.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      final String propertyName = e.getMessage().substring(startIndex + 1, endIndex);
+      errorHandler.onMissingParam(propertyName);
+      return null;
+    } catch (JsonParseException e) {
+      errorHandler.onInvalidJson();
+      return null;
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to deserialize buffer", e);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/PojoCodec.java]: ack

File: ksql-api/src/main/java/io/confluent/ksql/api/server/InsertsBodyHandler.java [hunk 8]
```diff
@@ -71,43 +72,37 @@ public void handleBodyEnd(final Void v) {
...
       } catch (DecodeException e) {
-        final JsonObject errResponse = ServerUtils
-            .createErrResponse(ErrorCodes.ERROR_CODE_INVALID_JSON,
-                "Invalid JSON in inserts stream");
-        routingContext.response().write(errResponse.toBuffer().appendString("\n")).end();
+        final ErrorResponse errorResponse = new ErrorResponse(
+            ErrorCodes.ERROR_CODE_MALFORMED_REQUEST,
+            "Invalid JSON in inserts stream");
+        routingContext.response().write(PojoCodec.serializeObject(errorResponse).appendString("\n"))
+            .end();
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/InsertsBodyHandler.java]: I've already split things out in a further PR, could introduce something like that there.

File: ksql-api/src/main/java/io/confluent/ksql/api/server/ServerUtils.java [hunk 9]
```diff
@@ -37,30 +37,44 @@ private ServerUtils() {
...
+      final Class<T> clazz) {
+    return PojoCodec.deserialiseObject(buffer, new HttpResponseErrorHandler(response), clazz);
+  }
+
+  private static class HttpResponseErrorHandler implements PojoDeserializerErrorHandler {
+
+    private final HttpServerResponse response;
+
+    HttpResponseErrorHandler(final HttpServerResponse response) {
+      this.response = response;
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/ServerUtils.java]: ack

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/InsertsStreamArgs.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.Map;
+
+/**
+ * Represents the arguments to an insert stream request
+ */
+public class InsertsStreamArgs {
+
+  public final String target;
+  public final boolean requiresAcks;
+  public final JsonObject properties;
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/InsertsStreamArgs.java]: I prefer to use JsonObject as it contains typesafe getters - e.g. getString, getInt etc

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/InsertResponse.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Represents a response to an insert
+ */
+public class InsertResponse {
+
+  public final String status;
+  public final Integer errorCode;
+  public final String message;
+
+  public InsertResponse(final Integer errorCode, final String message) {
```
[the remark this answers is not in the document]
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/InsertResponse.java]: ack

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamArgs.java (hunk 2 above)
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamArgs.java]: ack

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java (hunk 3 above)
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java]: ack

File: ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java (hunk 4 above)
Comment (purplefox) [reply ksql-api/src/main/java/io/confluent/ksql/api/server/protocol/QueryStreamResponseMetadata.java]: ack

**Pull request merged**